File: SwiftAssignments/SwiftAssignments/Questions/K06.swift
import Foundation

class K06 : Question {
    func question1() {
        let ages = getAges(number: 5)
        let count = ages.count
        let totalAges = ages.reduce(0.0, { acc, x in acc + x }) / Float(count)
        print("\(count)人の平均年齢は\(totalAges)です。")
    }

    func question2() {
        let upperPyramid = makeUpperPyramid(levels: 8).joined(separator: "\n")
        print(upperPyramid)
        print("")
        let lowerPyramid = makeLowerPyramid(levels: 8).joined(separator: "\n")
        print(lowerPyramid)
        print("")
        let specialPyramid = makeSpecialPyramid(levels: 8).joined(separator: "\n")
        print(specialPyramid)
        print("")
    }

    func question3() {
        let combinations = countHundreds(remains: 370).map { (h, f, t) in
            "10円の硬貨\(t)枚 50円の硬貨\(f)枚 100円の硬貨\(h)枚"
        }
        let count = combinations.count
        print(combinations.joined(separator: "\n") + "\n")
        print("以上\(count)通りを発見しました。")
    }

    func question4() {
        print("\t|\t", terminator: "")
        let numbers: [String] = Array(1...9).map { (i: Int) in String(i) }
        print(numbers.joined(separator: "\t"))
        print(String(repeating: "-", count: 90))

        var calculations: [[Int]] = []
        for i in 1...9 {
            var multiplications: [Int] = []
            for j in 1...9 {
                multiplications.append(i * j)
            }
            calculations.append(multiplications)
        }

        let result: [String] = calculations
                .enumerated()
                .map { (index: Int, nums: [Int]) in
                    let innerList = nums.map { i in String(i) }.joined(separator: "\t")
                    return "\(index + 1)\t|\t\(innerList)"
                }

        print(result.joined(separator: "\n"))
    }

    private func getAges(number: Int) -> [Float] {
        var numbers: [Float] = []
        for i in 0..<number {
            print("\(i + 1)人目の年齢を入力して下さい：")
            numbers.append(readLine().flatMap { s in Float(s) } ?? 0)
        }
        return numbers
    }

    private func makeUpperPyramid(levels: Int) -> [String] {
        var stars: [String] = []
        for i in 0..<levels {
            stars.append(String(repeating: "*", count: i + 1))
        }
        return stars
    }

    private func makeLowerPyramid(levels: Int) -> [String] {
        var stars: [String] = []
        for i in stride(from: levels, through: 1, by: -1) {
            stars.append(String(repeating: "*", count: i))
        }
        return stars
    }

    private func makeSpecialPyramid(levels: Int) -> [String] {
        var stars: [String] = []
        for i in 0..<levels {
            let str = String(repeating: " ", count: 8 - (i + 1)) + String(repeating: "*", count: i + 1)
            stars.append(str)
        }
        return stars
    }

    private func countTens(remains: Int) -> Int {
        remains / 10
    }

    private func countFifties(remains: Int) -> [(Int, Int)] {
        let max = remains / 50
        var combinations: [(Int, Int)] = []
        for i in stride(from: max, through: 0, by: -1) {
            let remaining = remains - (50 * i)
            combinations.append((i, countTens(remains: remaining)))
        }
        return combinations
    }

    private func countHundreds(remains: Int) -> [(Int, Int, Int)] {
        let max = remains / 100
        var combinations: [(Int, Int, Int)] = []
        for i in stride(from: max, through: 0, by: -1) {
            let remaining = remains - (100 * i)
            let combinationsWithHundred = countFifties(remains: remaining).map { (f, t) in (i, f, t) }
            combinations.append(contentsOf: combinationsWithHundred)
        }
        return combinations
    }

}

File: GoRewrite/questions/k06.go
package questions

import (
	"fmt"
	"strings"
)

type K06 struct {}

func (k K06) question1() {
	var ages []int
	for i := 0; i < 5; i++ {
		fmt.Printf("%d人目の年齢を入力して下さい：", i + 1)
		var input int
		_, _ = fmt.Scanf("%d", &input)
		ages = append(ages, input)
	}
	count := len(ages)
	var totalAges float32
	for _, age := range ages {
		totalAges += float32(age)
	}
	fmt.Printf("%d人の平均年齢は%fです。", count, totalAges / float32(count))
}

func (k K06) question2() {
	for i := 0; i < 8; i++ {
		for j := 0; j < i + 1; j++ {
			fmt.Print("*")
		}
		fmt.Println()
	}
	fmt.Println()

	for i := 8; i > 0; i-- {
		for j := 0; j < i; j++ {
			fmt.Print("*")
		}
		fmt.Println()
	}
	fmt.Println()

	for i := 8; i > 0; i-- {
		for j := 0; j < i - 1; j++ {
			fmt.Print(" ")
		}

		for k := 8; k > i - 1; k-- {
			fmt.Print("*")
		}
		fmt.Println()
	}
	fmt.Println()
}

func (k K06) question3() {
	total := 370
	count := 0

	for i := 0; i <= total / 100; i++ {
		for j := 0; j <= total / 50; j++ {
			for k := 0; k <= total / 10; k++ {
				if i * 100 + j * 50 + k * 10 != total {
					continue
				}
				fmt.Printf("10円の硬貨%d枚 50円の硬貨%d枚 100円の硬貨%d枚\n", k, j, i)
				count += 1
			}
		}
	}
	fmt.Printf("\n以上%d通りを発見しました。", count)
}

func (k K06) question4() {
	fmt.Print("\t|\t")
	for i := 1; i < 10; i++ {
		fmt.Printf("%d\t", i)
	}
	fmt.Println()
	fmt.Println(strings.Repeat("-", 90))
	for i := 1; i < 10; i++ {
		fmt.Printf("%d\t|\t", i)
		for j := 1; j < 10; j++ {
			fmt.Printf("%d\t", i * j)
		}
		fmt.Println()
	}
}

func (k K06) Execute(number int) {
	switch number {
	case 1:
		k.question1()
	case 2:
		k.question2()
	case 3:
		k.question3()
	case 4:
		k.question4()
	}
}